### PR TITLE
FolkLSTM PR

### PR DESCRIPTION
**Pull request by WarmBuns**
The slightly massive PR that includes our first model + all the code for preprocessing songs for usage with RNN/LSTM/Time Series models + converting between the time series representations back to MIDI, along with some placeholder code for future issues

### CODE REVIEW

File: src/common/melody_generator.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+        return melody
+
+
+    def _sample_with_temperature(self, probabilities, temperature=0.7):
+        """ 
+        The temperature modifies the probability distribution
+        the lower the temperature, the more deterministic the sampling is
+        if the temperature is 1 that is just default settings
+        higher temperature, more exploritive model
```
Comment (lloydoad) [inline src/common/melody_generator.py]: Adding the datatypes for the parameters may also be helpful to others adding to this.

File: src/common/melody_generator.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        """
+        predicitions = np.log(probabilities) / temperature
```
Comment (lloydoad) [inline src/common/melody_generator.py:94]: Nit: `predicitions`

File: src/common/melody_generator.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+        return index
+
+    def save_melody(self, melody, file_name, format="midi", step_duration=0.25):
+        """ 
+            * The melody object you want to pass
+            * the name of the saved file
+            * the file format you want to save it in
+            * The step duration you trained the model in, 0.25 is a sixteenth note
+        """
```
Comment (lloydoad) [inline src/common/melody_generator.py]: documentation format is inconsistent

File: src/common/melody_generator.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                    if start_symbol == "r":
+                        m21_event = m21.note.Rest(quarterLength=quarter_length_duration)
+                    else:
+                        m21_event = m21.note.Note(int(start_symbol), quarterLength=quarter_length_duration)
+
+                    stream.append(m21_event)
+                    
+                    # reset to the other note
+                    step_counter = 1
+                start_symbol = symbol
```
Comment (lloydoad) [inline src/common/melody_generator.py:141]: Consider: A guard structure instead of nested if-blocks. For example 
```
if (badCondition1) {
    break
}

if (badCondition2) {
    break
}

foobar()
```

instead of if-blocks inside of if-blocks. The former makes it clearer what we don't want and easier to debug for someone who didn't create the original codebase

File: src/common/melody_generator.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        # write the m21 stream to midi file
+        stream.write(format, file_name)
+
+
+
+
+
+
+
+
```
Comment (lloydoad) [inline src/common/melody_generator.py]: Nit: Unnecessary newlines

File: src/common/music_helper.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+                    encoded_song.append("_")
+
+        # cast encoded song to str
+        encoded_song = " ".join(map(str, encoded_song))
+
+        return encoded_song
+
+    def convert_songs_to_int(self, songs, mapping_path:str):
+        """ 
+
```
Comment (lloydoad) [inline src/common/music_helper.py:141]: Nit: spacing

File: src/common/music_helper.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+            mappings[symbol] = i
+
+        # save voabulary to a json file
+        # TODO: Have the file_helper do this
+        with open(mapping_path, "w+") as fp:
+            json.dump(mappings, fp, indent=4)
+
+    def generate_training_sequences(self, sequence_length, single_file_dataset, mapping_path):
+        """ 
+
```
Comment (lloydoad) [inline src/common/music_helper.py]: Missing docs

**Review by lloydoad — COMMENTED** [review]
It's looking good. Comments are mostly about formatting and structure. Looks solid logic-wise but will require someone with more ML knowledge to approve. Still recommend combing through and looking at documentation consistency, newline consistency and a few spelling errors.

File: src/common/melody_generator.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        # map seed to int
+        seed = [self._mappings[symbol] for symbol in seed]
+
+        for _ in range(num_steps):
+            # limit the seed to max_seq_len
+            seed = seed[-max_seq_len:] 
+
+            # ont-hot encode the seed
+            onehot_seed = keras.utils.to_categorical(seed, num_classes=len(self._mappings))
```
Comment (jcbang) [inline src/common/melody_generator.py:58]: What's the cardinality on the seed values? Not something that would require revision, but is helpful to know for curiosity sake

File: src/common/melody_generator.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            self._mappings = json.load(fp)
+
+    
+    def generate_melody(self, seed, num_steps, max_seq_len, temperature):
+        """ 
+        * seed - what we want to pass to the network and the network countinues that seed
+        * num_steps - How far we want the network to predict
+        * max_seq_len - How many steps of the seeds do we want to consider for the network (the seed will grow large), 
+        so we pass the seed as a sliding window
+        """
```
Comment (jcbang) [inline src/common/melody_generator.py:41]: Consider using the format:
```
Parameters
------------
stuff - stuff

Returns
--------
stuff - stuff
```
to standardize comments about what the inputs and outputs are, adding underlines helps text editors with hover-over header formatting

**Comment by jcbang** [comment]
+1 on Lloyd's suggestions with consistent styling before merge

File: src/common/melody_generator.py (hunk 8 above)
Comment (WarmBuns) [reply src/common/melody_generator.py:58]: @jcbang That is dependent on your data, if your data is diverse and has a lot of notes your cardinality is higher, that is the purpose of _mappings, Adding documentation to explain that

File: src/common/melody_generator.py (hunk 4 above)
Comment (WarmBuns) [reply src/common/melody_generator.py:141]: This is a good suggestion, I actually want to refactor a lot of melody_generator so I am going to leave this change for the next issue involving that refactor (since technically its not an all purpose melody_generator, its only applicable to LSTMs networks), I'll reference this as a requirement for that issue

**Review by jcbang — DISMISSED** [review]
LGTM
